File: cpp/src/msgpack/rpc/session.cc
//
// msgpack::rpc::session - MessagePack-RPC for C++
//
// Copyright (C) 2009-2010 FURUHASHI Sadayuki
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.
//
#include "session.h"
#include "session_impl.h"
#include "future_impl.h"
#include "request_impl.h"
#include "exception_impl.h"
#include "cclog/cclog.h"

namespace msgpack {
namespace rpc {


session_impl::session_impl(const address& addr, loop lo) :
	m_addr(addr),
	m_loop(lo),
	m_msgid_rr(0),  // FIXME rand()?
	m_timeout(30)
{ }

session_impl::~session_impl() { }

void session_impl::build(const builder& b)
{
	m_tran = b.build(this, m_addr);
	m_timeout = b.get_timeout();
}

shared_session session_impl::create(const builder& b, const address addr, loop lo)
{
	shared_session s(new session_impl(addr, lo));
	s->build(b);
	return s;
}

future session_impl::send_request_impl(msgid_t msgid, sbuffer* sbuf)
{
	LOG_DEBUG("sending... msgid=",msgid);
	shared_future f(new future_impl(shared_from_this(), m_loop));
	m_reqtable.insert(msgid, f);

	m_tran->send_data(sbuf);

	return future(f);
}

future session_impl::send_request_impl(msgid_t msgid, auto_vreflife vbuf)
{
	LOG_DEBUG("sending... msgid=",msgid);
	shared_future f(new future_impl(shared_from_this(), m_loop));
	m_reqtable.insert(msgid, f);

	m_tran->send_data(vbuf);

	return future(f);
}

void session_impl::send_notify_impl(sbuffer* sbuf)
{
	m_tran->send_data(sbuf);
}

void session_impl::send_notify_impl(auto_vreflife vbuf)
{
	m_tran->send_data(vbuf);
}

msgid_t session_impl::next_msgid()
{
	// FIXME __sync_add_and_fetch
	return __sync_add_and_fetch(&m_msgid_rr, 1);
}


void session_impl::step_timeout()
{
	LOG_TRACE("step_timeout");
	std::vector<shared_future> timedout;
	m_reqtable.step_timeout(&timedout);
	if(timedout.empty()) { return ;}
	for(std::vector<shared_future>::iterator it(timedout.begin()),
			it_end(timedout.end()); it != it_end; ++it) {
		shared_future& f = *it;
		f->set_result(object(), TIMEOUT_ERROR, auto_zone());
	}
}

void session_impl::on_connect_failed()
{
	std::vector<shared_future> all;
	m_reqtable.take_all(&all);
	for(std::vector<shared_future>::iterator it(all.begin()),
			it_end(all.end()); it != it_end; ++it) {
		shared_future& f = *it;
		f->set_result(object(), CONNECT_ERROR, auto_zone());
	}
}

void session_impl::on_response(msgid_t msgid,
		object result, object error, auto_zone z)
{
	LOG_TRACE("response result=",result," error=",error);
	shared_future f = m_reqtable.take(msgid);
	if(!f) {
		LOG_DEBUG("no entry on request table for msgid=",msgid);
		return;
	}
	f->set_result(result, error, z);
}


const address& session::get_address() const
	{ return m_pimpl->get_address(); }

const loop& session::get_loop() const
	{ return const_cast<const session_impl*>(m_pimpl.get())->get_loop(); }

loop session::get_loop()
	{ return m_pimpl->get_loop(); }

void session::set_timeout(unsigned int sec)
	{ m_pimpl->set_timeout(sec); }

unsigned int session::get_timeout() const
	{ return m_pimpl->get_timeout(); }

future session::send_request_impl(msgid_t msgid, std::auto_ptr<with_shared_zone<vrefbuffer> > vbuf)
	{ return m_pimpl->send_request_impl(msgid, vbuf); }

future session::send_request_impl(msgid_t msgid, sbuffer* sbuf)
	{ return m_pimpl->send_request_impl(msgid, sbuf); }

void session::send_notify_impl(sbuffer* sbuf)
	{ return m_pimpl->send_notify_impl(sbuf); }

void session::send_notify_impl(std::auto_ptr<with_shared_zone<vrefbuffer> > vbuf)
	{ return m_pimpl->send_notify_impl(vbuf); }

msgid_t session::next_msgid()
	{ return m_pimpl->next_msgid(); }


}  // namespace rpc
}  // namespace msgpack

File: ruby/lib/msgpack/rpc/session.rb
#
# MessagePack-RPC for Ruby
#
# Copyright (C) 2010-2011 FURUHASHI Sadayuki
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.
#
module MessagePack
module RPC


# Session is a abstract class corresponds with a remote host.
# You can call remote method using call, call_async, callback or notify method.
class Session
	def initialize(builder, address, loop)
		@address = address
		@loop = loop
		@reqtable = {}
		@timeout = 10    # FIXME default timeout time
		@seqid = 0
		@transport = builder.build_transport(self, address)
	end
	attr_reader :loop, :address

	# Sets and gets timeout in seconds.
	attr_accessor :timeout

	# backward compatibility
	def port  #:nodoc:
		@address.port
	end

	# backward compatibility
	def host  #:nodoc:
		@address.host
	end

	# call-seq:
	#   call(symbol, *args) -> result of remote method
	#
	# Calls remote method.
	# This method is same as call_async(method, *args).get
	def call(method, *args)
		send_request(method, args).get
	end

	# call-seq:
	#   call_apply(symbol, params) -> result of remote method
	#
	# Calls remote method.
	# This method is same as call(method, *args) excepting that
	# the arugment is a array.
	def call_apply(method, params)
		send_request(method, params).get
	end

	# call-seq:
	#   call_async(symbol, *args) -> Future
	#
	# Calls remote method asynchronously.
	# This method is non-blocking and returns Future.
	def call_async(method, *args)
		future = send_request(method, args)
	end

	# call-seq:
	#   call_async_apply(symbol, params) -> Future
	#
	# Calls remote method asynchronously.
	# This method is same as call_async(method, *args) excepting that
	# the arugment is a array.
	def call_async_apply(method, params)
		future = send_request(method, params)
	end

	# backward compatibility
	alias_method :send_without_call_async, :send
	def send(method, *args)
	    if caller.first =~ /.*_test.rb/ || caller.first =~ /.*_spec.rb/ then
	    	    warn "\n Don't use send method. Use 'call_async' method."
            end
	    call_async(method, *args)
	end


	# call-seq:
	#   callback(symbol, *args) {|future| }
	#
	# Calls remote method asynchronously.
	# The callback method is called with Future when the result is reached.
	# This method is same as call_async(method, *args).attach_callback {|future|  }
	def callback(method, *args, &block)
		future = send_request(method, args)
		future.attach_callback(block)
		future
	end

	# call-seq:
	#   callback_apply(symbol, params) {|future| }
	#
	# Calls remote method asynchronously.
	# The callback method is called with Future when the result is reached.
	# This method is same as callback(method, *args).attach_callback {|future|  }
	# excepting that the argument is a array.
	def callback_apply(method, params, &block)
		future = send_request(method, params)
		future.attach_callback(block)
		future
	end

	# call-seq:
	#   notify(symbol, *args) -> nil
	#
	# Calls remote method with NOTIFY protocol.
	# It doesn't require server to return results.
	# This method is non-blocking and returns nil.
	def notify(method, *args)
		send_notify(method, args)
		nil
	end

	# call-seq:
	#   notify_apply(symbol, params) -> nil
	#
	# Calls remote method with NOTIFY protocol.
	# It doesn't require server to return results.
	# This method is non-blocking and returns nil.
	# This method is same as notify(method, *args) excepting that
	# the argument is a array.
	def notify_apply(method, params)
		send_notify(method, params)
		nil
	end

	# Closes underlaying Transport and destroy resources.
	def close
		@transport.close
		@reqtable = {}
		@seqid = 0
		self
	end

	# from ClientTransport
	def on_response(sock, msgid, error, result)  #:nodoc:
		if future = @reqtable.delete(msgid)
			future.set_result(error, result)
		end
	end

	# from ClientTransport
	def on_connect_failed  #:nodoc:
		@reqtable.reject! {|msgid, future|
			future.set_result ConnectionTimeoutError::CODE, ["connection timed out"]
			true
		}
		nil
	end

	# from Client, SessionPool
	def step_timeout  #:nodoc:
		timedout = []
		@reqtable.reject! {|msgid, future|
			if future.step_timeout
				timedout.push(future)
				true
			end
		}
		timedout.each {|future|
			future.set_result TimeoutError::CODE, ["request timed out"]
		}
		!@reqtable.empty?
	end

	private
	def send_request(method, param)
		method = method.to_s
		msgid = @seqid
		@seqid += 1; if @seqid >= 1<<31 then @seqid = 0 end
		data = [REQUEST, msgid, method, param].to_msgpack
		@transport.send_data(data)
		@reqtable[msgid] = Future.new(self, @loop)
	end

	def send_notify(method, param)
		method = method.to_s
		data = [NOTIFY, method, param].to_msgpack
		@transport.send_data(data)
		nil
	end
end


end
end
